### PR TITLE
More updates

### PR DESCRIPTION
**Pull request by xiehaom7**
Added numbers.Log10; 
Added Iterators for object ans string; 
Added su.Iter and su.Next for "for...in" statement;
Added calling string;
Added Object.Join

### CODE REVIEW

File: runtime/builtin/objects.ts [hunk 1]
```diff
@@ -186,3 +186,64 @@ export function su_objectq(x: any = mandatory()): boolean {
...
+(SuObject.prototype['Iter'] as any).$callNamed = function (_named: any) {
+    return SuObject.prototype['Iter'].call(this);
+};
+(SuObject.prototype['Iter'] as any).$callAt = function (args: SuObject) {
+    return (SuObject.prototype['Iter'] as any).$callNamed.call(this, util.mapToOb(args.map), ...args.vec);
+};
+(SuObject.prototype['Iter'] as any).$params = '';
+//GENERATED end
+
+//BUILTIN SuObject.Join(seperator="")
```
Comment (apmckinlay) [inline runtime/builtin/objects.ts]: Spelling mistake - should be "separator"

File: runtime/su.ts [hunk 2]
```diff
@@ -282,32 +282,39 @@ export function mkdate(s: string): SuDate {
     return SuDate.literal(s) !;
 }
 
+enum CallType {call, callNamed, callAt}
 // Note: only Suneido values and strings are callable
-
+export let call = callFnGenerator(CallType.call);
+export let callNamed = callFnGenerator(CallType.callNamed);
+export let callAt = callFnGenerator(CallType.callAt);
 // Note: using apply until better spread (...) performance in v8
```
Comment (apmckinlay) [inline runtime/su.ts]: I think spread performance has been improved

File: runtime/su.ts [hunk 3]
```diff
@@ -282,32 +282,39 @@ export function mkdate(s: string): SuDate {
...
-    cantCall(f);
-}
-
-export function callAt(f: any, args: SuObject): any {
-    let call = f.$callAt;
-    if (typeof call === 'function')
-        return call(args);
-    //  TODO strings
-    cantCall(f);
+function callFnGenerator(type: CallType):
```
Comment (apmckinlay) [inline runtime/su.ts]: Might be worth checking how the speed of this approach compares to writing out the three functions. I think for this code speed is more important than duplication.

File: runtime/suobject.ts [hunk 4]
```diff
@@ -93,13 +94,21 @@ export class SuObject extends SuValue {
     Add(x: any = mandatory()): SuObject {
         maxargs(1, arguments.length);
         this.checkReadonly();
+        return this.runWithModificationCheck(this.addFn, x);
+    }
+
+    private addFn(x: any): SuObject {
```
Comment (apmckinlay) [inline runtime/suobject.ts]: Unless we need addFn, putFn, etc. for other reasons, we can just put them inline (probably in an arrow function), like we would with a block in Suneido. Then you don't need to pass any arguments.

File: runtime/suobject.ts [hunk 5]
```diff
@@ -298,8 +315,82 @@ export class SuObject extends SuValue {
...
+    private runWithModificationCheck<T>(fn: (...args: any[]) => T, ...args: any[]): T {
+        let vecSize = this.vecsize();
+        let mapSize = this.mapsize();
+        let res = fn.apply(this, args);
+        if (vecSize !== this.vecsize() || mapSize !== this.mapsize())
+            this.version++;
+        return res;
+    }
+
+    Join(seperator: string = ''): string {
```
Comment (apmckinlay) [inline runtime/suobject.ts]: Should be "separator"

File: runtime/su.ts (hunk 3 above)
Comment (apmckinlay) [reply runtime/su.ts]: Also better to do as much work as possible outside the generated function.  e.g. '$' + CallType[type]

**Comment by xiehaom7** [comment]
Fixed the aforementioned issues

**Pull request merged**